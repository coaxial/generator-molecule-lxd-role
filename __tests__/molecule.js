'use strict';
const { readFileSync } = require('fs');
const assert = require('yeoman-assert');
const helpers = require('yeoman-test');
const path = require('path');

describe('generator-molecule-lxd-role:molecule', () => {
  const subgeneratorPath = '../generators/molecule';
  const modes = ['role', 'playbook'];

  modes.forEach(mode => {
    describe(`when generating a ${mode}`, () => {
      const defaultResponses = {};
      const defaultOptions = {
        repoName: 'test-role',
        projectName: 'Test project name',
        targetDistributions: ['UBUNTU'],
        mode,
        useTravis: true,
        targetVersions: [
          {
            family: 'debian',
            distribution: 'ubuntu',
            codeName: 'trusty',
            versionNumber: '14.04',
            tags: ['lts', 'current'],
          },
          {
            family: 'debian',
            distribution: 'ubuntu',
            codeName: 'xenial',
            versionNumber: '16.04',
            tags: ['lts', 'current'],
          },
          {
            family: 'debian',
            distribution: 'ubuntu',
            codeName: 'bionic',
            versionNumber: '18.04',
            tags: ['lts', 'current'],
          },
        ],
      };

      describe('when Travis disabled', () => {
        beforeAll(() => {
          return helpers
            .run(path.join(__dirname, subgeneratorPath))
            .withOptions({ ...defaultOptions, useTravis: false })
            .withPrompts(defaultResponses);
        });

        describe('.travis.yml', () => {
          const filePath = '.travis.yml';

          it('does not exist', () => {
            assert.noFile(filePath);
          });
        });

        describe('.travis/setup.sh', () => {
          const filePath = '.travis/setup.sh';

          it('does not exist', () => {
            assert.noFile(filePath);
          });
        });
      });

      describe('when all the prompts have answers', () => {
        beforeAll(() => {
          return helpers
            .run(path.join(__dirname, subgeneratorPath))
            .withOptions({
              ...defaultOptions,
              requirements: [
                { src: 'user/test' },
                {
                  name: 'test',
                  src: 'https://bithub.com/user/test.git',
                },
              ],
            })
            .withPrompts(defaultResponses);
        });

        if (mode === 'playbook') {
          describe('requirements.yml', () => {
            const filePath = 'requirements.yml';

            it('exists', () => {
              assert.file(filePath);
            });

            it('is correctly formatted', () => {
              const actual = readFileSync(filePath, 'utf8');

              expect(actual).toMatchSnapshot();
            });
          });
        }

        if (mode === 'role') {
          describe('requirements.yml', () => {
            const filePath = 'requirements.yml';

            it('does not exists', () => {
              assert.noFile(filePath);
            });
          });
        }

        describe('molecule/default/requirements.yml', () => {
          const filePath = 'molecule/default/requirements.yml';

          it('exists', () => {
            assert.file(filePath);
          });

          it('is correctly formatted', () => {
            const actual = readFileSync(filePath, 'utf8');

            expect(actual).toMatchSnapshot();
          });
        });

        describe('.yamllint', () => {
          const filePath = '.yamllint';

          it('exists', () => {
            assert.file(filePath);
          });

          it('is correctly formatted', () => {
            const actual = readFileSync(filePath, 'utf8');

            expect(actual).toMatchSnapshot();
          });
        });

        describe('.travis/setup.sh', () => {
          const filePath = '.travis/setup.sh';

          it('exists', () => {
            assert.file(filePath);
          });

          it('is correctly formatted', () => {
            const actual = readFileSync(filePath, 'utf8');

            expect(actual).toMatchSnapshot();
          });
        });

        describe('.travis.yml', () => {
          const filePath = '.travis.yml';

          it('exists', () => {
            assert.file(filePath);
          });

          it('is correctly formatted', () => {
            const actual = readFileSync(filePath, 'utf8');

            expect(actual).toMatchSnapshot();
          });
        });

        describe('molecule/create.yml', () => {
          const filePath = 'molecule/create.yml';

          it('exists', () => {
            assert.file(filePath);
          });

          it('is correctly formatted', () => {
            const actual = readFileSync(filePath, 'utf8');

            expect(actual).toMatchSnapshot();
          });
        });

        describe('molecule/destroy.yml', () => {
          const filePath = 'molecule/destroy.yml';

          it('exists', () => {
            assert.file(filePath);
          });

          it('is correctly formatted', () => {
            const actual = readFileSync(filePath, 'utf8');

            expect(actual).toMatchSnapshot();
          });
        });

        describe('molecule/default/molecule.yml', () => {
          const filePath = 'molecule/default/molecule.yml';

          it('exists', () => {
            assert.file(filePath);
          });

          it('is correctly formatted', () => {
            const actual = readFileSync(filePath, 'utf8');

            expect(actual).toMatchSnapshot();
          });
        });

        describe('molecule/default/playbook.yml', () => {
          const filePath = 'molecule/default/playbook.yml';

          it('exists', () => {
            assert.file(filePath);
          });

          it('is correctly formatted', () => {
            const actual = readFileSync(filePath, 'utf8');

            expect(actual).toMatchSnapshot();
          });
        });

        describe('molecule/default/tests/test_default.py', () => {
          const filePath = 'molecule/default/tests/test_default.py';

          it('exists', () => {
            assert.file(filePath);
          });

          it('is correctly formatted', () => {
            const actual = readFileSync(filePath, 'utf8');

            expect(actual).toMatchSnapshot();
          });
        });

        describe('.yamllint', () => {
          const filePath = '.yamllint';

          it('exists', () => {
            assert.file(filePath);
          });

          it('is correctly formatted', () => {
            const actual = readFileSync(filePath, 'utf8');

            expect(actual).toMatchSnapshot();
          });
        });

        describe('.gitignore', () => {
          const filePath = '.gitignore';

          it('exists', () => {
            assert.file(filePath);
          });

          it('is correctly formatted', () => {
            const actual = readFileSync(filePath, 'utf8');

            expect(actual).toMatchSnapshot();
          });
        });
      });

      describe('when targetting another distribution than Ubuntu', () => {
        beforeAll(() => {
          return helpers
            .run(path.join(__dirname, subgeneratorPath))
            .withOptions({
              ...defaultOptions,
              targetVersions: [
                {
                  family: 'debian',
                  distribution: 'debian',
                  codeName: 'jessie',
                  versionNumber: '8',
                  tags: ['current'],
                },
                {
                  family: 'debian',
                  distribution: 'debian',
                  codeName: 'stretch',
                  versionNumber: '9',
                  tags: ['lts'],
                },
              ],
              targetDistributions: ['DEBIAN'],
            })
            .withPrompts({
              ...defaultResponses,
            });
        });

        if (mode === 'playbook') {
          describe('requirements.yml', () => {
            const filePath = 'requirements.yml';

            it('exists', () => {
              assert.file(filePath);
            });

            it('is correctly formatted', () => {
              const actual = readFileSync(filePath, 'utf8');

              expect(actual).toMatchSnapshot();
            });
          });
        }

        if (mode === 'role') {
          describe('requirements.yml', () => {
            const filePath = 'requirements.yml';

            it('does not exists', () => {
              assert.noFile(filePath);
            });
          });
        }

        describe('molecule/default/requirements.yml', () => {
          const filePath = 'molecule/default/requirements.yml';

          it('exists', () => {
            assert.file(filePath);
          });

          it('is correctly formatted', () => {
            const actual = readFileSync(filePath, 'utf8');

            expect(actual).toMatchSnapshot();
          });
        });

        describe('molecule/default/molecule.yml', () => {
          const filePath = 'molecule/default/molecule.yml';

          it('formats the platform list properly', () => {
            const actual = readFileSync(filePath, 'utf8');

            expect(actual).toMatchSnapshot();
          });
        });
      });
    });
  });
});
