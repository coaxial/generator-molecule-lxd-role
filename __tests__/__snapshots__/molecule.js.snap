// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`generator-molecule-lxd-role:molecule when all the prompts have answers .gitignore is correctly formatted 1`] = `
"*.pyc
*.log
"
`;

exports[`generator-molecule-lxd-role:molecule when all the prompts have answers .travis.yml is correctly formatted 1`] = `
"---
jobs:
  include:
    - &test
      stage: test
      language: python
      python: \\"2.7\\"
      cache: pip
      dist: xenial
      sudo: required
      install:
        - .travis/setup.sh
      env:
        # keep fact gathering from timing out
        - ANSIBLE_GATHER_TIMEOUT=300
      script:
        # the python venv needs to be in the path for molecule to run
        # https://unix.stackexchange.com/questions/83191/how-to-make-sudo-preserve-path
        - sudo env \\"PATH=$PATH\\" \\"ANSIBLE_GATHER_TIMEOUT=$ANSIBLE_GATHER_TIMEOUT\\" molecule test

notifications:
  webhooks: https://galaxy.ansible.com/api/v1/notifications/
"
`;

exports[`generator-molecule-lxd-role:molecule when all the prompts have answers .travis/setup.sh is correctly formatted 1`] = `
"#!/bin/sh
sudo apt-get update
# xenial ships LXD 2 LTS, backports ships 3 LTS
sudo apt-get upgrade -t xenial-backports -y lxd lxd-client

sudo lxd --version
# Wait for the socket to be ready
sudo lxd waitready
sudo lxd init --auto

pip install ansible
pip install molecule

# Enable nesting without privileged containers,
# cf. https://stgraber.org/2017/06/15/custom-user-mappings-in-lxd-containers/
sudo sed -i 's/^\\\\(lxd:[0-9]\\\\{1,\\\\}\\\\):.*/\\\\1:1000000000/' /etc/subuid
sudo sed -i 's/^\\\\(root:[0-9]\\\\{1,\\\\}\\\\):.*/\\\\1:1000000000/' /etc/subuid
sudo sed -i 's/^\\\\(lxd:[0-9]\\\\{1,\\\\}\\\\):.*/\\\\1:1000000000/' /etc/subgid
sudo sed -i 's/^\\\\(root:[0-9]\\\\{1,\\\\}\\\\):.*/\\\\1:1000000000/' /etc/subgid
sudo systemctl restart lxd
"
`;

exports[`generator-molecule-lxd-role:molecule when all the prompts have answers .yamllint is correctly formatted 1`] = `
"extends: default

rules:
  braces:
    max-spaces-inside: 1
    level: error
  brackets:
    max-spaces-inside: 1
    level: error
  line-length: disable
"
`;

exports[`generator-molecule-lxd-role:molecule when all the prompts have answers molecule/create.yml is correctly formatted 1`] = `
"---
- name: Create
  hosts: localhost
  connection: local
  gather_facts: false
  no_log: \\"{{ not lookup('env', 'MOLECULE_DEBUG') | bool }}\\"
  tasks:
    - name: Create molecule instance(s)
      lxd_container:
        name: \\"{{ item.name }}\\"
        state: started
        source:
          type: image
          mode: pull
          server: https://images.linuxcontainers.org
          protocol: simplestreams
          alias: \\"{{ item.alias }}/amd64\\"
        profiles: [\\"default\\"]
        wait_for_ipv4_addresses: true
        timeout: 600
      with_items: \\"{{ molecule_yml.platforms }}\\"

    - name: Install Python in container
      delegate_to: \\"{{ item.name }}\\"
      raw: apt-get install -y python
      with_items: \\"{{ molecule_yml.platforms }}\\"
"
`;

exports[`generator-molecule-lxd-role:molecule when all the prompts have answers molecule/default/molecule.yml is correctly formatted 1`] = `
"---
dependency:
  name: galaxy
driver:
  name: lxd
lint:
  name: yamllint
platforms:
  - name: ubuntu-trusty
    alias: ubuntu/trusty
  - name: ubuntu-xenial
    alias: ubuntu/xenial
  - name: ubuntu-bionic
    alias: ubuntu/bionic

provisioner:
  name: ansible
  playbooks:
    create: ../create.yml
    destroy: ../destroy.yml
  lint:
    name: ansible-lint
scenario:
  name: default
verifier:
  name: testinfra
  lint:
    name: flake8
"
`;

exports[`generator-molecule-lxd-role:molecule when all the prompts have answers molecule/default/playbook.yml is correctly formatted 1`] = `
"---
- name: Converge
  hosts: all
  roles:
    - name: ansible-role-test-role
"
`;

exports[`generator-molecule-lxd-role:molecule when all the prompts have answers molecule/default/tests/test_default.py is correctly formatted 1`] = `
"import os

import testinfra.utils.ansible_runner

testinfra_hosts = testinfra.utils.ansible_runner.AnsibleRunner(
    os.environ['MOLECULE_INVENTORY_FILE']).get_hosts('all')


def test_example(host):
    file = host.file('/etc/hosts')

    assert file.exists
    assert file.user == 'root'
    assert file.group == 'root'
"
`;

exports[`generator-molecule-lxd-role:molecule when all the prompts have answers molecule/destroy.yml is correctly formatted 1`] = `
"---
- name: Destroy
  hosts: localhost
  connection: local
  gather_facts: false
  no_log: \\"{{ not lookup('env', 'MOLECULE_DEBUG') | bool }}\\"
  tasks:
    - name: Destroy molecule instance(s)
      lxd_container:
        name: \\"{{ item.name }}\\"
        state: absent
        force_stop: \\"{{ item.force_stop | default(true) }}\\"
      with_items: \\"{{ molecule_yml.platforms }}\\"
"
`;

exports[`generator-molecule-lxd-role:molecule when generating for a playbook molecule/default/molecule.yml is correctly formatted 1`] = `
"---
dependency:
  name: galaxy
driver:
  name: lxd
lint:
  name: yamllint
platforms:
  - name: ubuntu-trusty
    alias: ubuntu/trusty
  - name: ubuntu-xenial
    alias: ubuntu/xenial
  - name: ubuntu-bionic
    alias: ubuntu/bionic

provisioner:
  name: ansible
  playbooks:
    create: ../create.yml
    converge: ../../playbook.yml
    destroy: ../destroy.yml
  lint:
    name: ansible-lint
scenario:
  name: default
verifier:
  name: testinfra
  lint:
    name: flake8
"
`;

exports[`generator-molecule-lxd-role:molecule when generating for a role molecule/default/molecule.yml is correctly formatted 1`] = `
"---
dependency:
  name: galaxy
driver:
  name: lxd
lint:
  name: yamllint
platforms:
  - name: ubuntu-trusty
    alias: ubuntu/trusty
  - name: ubuntu-xenial
    alias: ubuntu/xenial
  - name: ubuntu-bionic
    alias: ubuntu/bionic

provisioner:
  name: ansible
  playbooks:
    create: ../create.yml
    destroy: ../destroy.yml
  lint:
    name: ansible-lint
scenario:
  name: default
verifier:
  name: testinfra
  lint:
    name: flake8
"
`;

exports[`generator-molecule-lxd-role:molecule when generating for a role molecule/default/playbook.yml is correctly formatted 1`] = `
"---
- name: Converge
  hosts: all
  roles:
    - name: ansible-role-test-role
"
`;

exports[`generator-molecule-lxd-role:molecule when targetting another distribution than Ubuntu molecule/default/molecule.yml formats the platform list properly 1`] = `
"---
dependency:
  name: galaxy
driver:
  name: lxd
lint:
  name: yamllint
platforms:
  - name: debian-jessie
    alias: debian/jessie
  - name: debian-stretch
    alias: debian/stretch

provisioner:
  name: ansible
  playbooks:
    create: ../create.yml
    destroy: ../destroy.yml
  lint:
    name: ansible-lint
scenario:
  name: default
verifier:
  name: testinfra
  lint:
    name: flake8
"
`;
